packages/loopring_v3/contracts/impl/AgentRegistry.sol

```diff
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.6.10;
+pragma experimental ABIEncoderV2;
+
+import "../iface/IAgentRegistry.sol";
+
+import "../lib/AddressSet.sol";
+import "../lib/Claimable.sol";
+
+
+contract AgentRegistry is IAgentRegistry, AddressSet, Claimable
+{
+    bytes32 internal constant UNIVERSAL_AGENTS = keccak256("__UNVERSAL_AGENTS__");
+
+    event AgentRegistered(
+        address user,
+        address agent,
+        bool    registered
+    );
+
+    constructor() public Claimable() {}
+
+    function isAgent(
+        address user,
+        address agent
+        )
+        external
+        override
+        view
+        returns (bool)
+    {
+        return isUnversalAgent(agent) || isUserAgent(user, agent);
+    }
+
+    function registerUniversalAgent(
+        address agent,
+        bool    toRegister
+        )
+        external
+        onlyOwner
+    {
+        registerInternal(UNIVERSAL_AGENTS, agent, toRegister);
+        emit AgentRegistered(address(0), agent, toRegister);
+    }
+
+    function getUniversalAgents()
+        public
+        view
+        returns (address[] memory)
+    {
+        return addressesInSet(UNIVERSAL_AGENTS);
+    }
+
+    function isUnversalAgent(address agent)
+        public
+        view
+        returns (bool)
+    {
+        return isAddressInSet(UNIVERSAL_AGENTS, agent);
+    }
+
+    function registerUserAgent(
+        address agent,
+        bool    toRegister
+        )
+        external
+    {
+        registerInternal(userKey(msg.sender), agent, toRegister);
+        emit AgentRegistered(msg.sender, agent, toRegister);
+    }
+
+    function getUserAgents(address user)
+        public
+        view
+        returns (address[] memory agents)
+    {
+        if (user != address(0)) {
+            agents = addressesInSet(userKey(user));
+        }
+    }
+
+    function isUserAgent(
+        address user,
+        address agent
+        )
+        public
+        view
+        returns (bool)
+    {
+        return isAddressInSet(userKey(user), agent);
+    }
+
+    function registerInternal(
+        bytes32 key,
+        address agent,
+        bool    toRegister
+        )
+        private
+    {
+        require(agent != address(0), "ZERO_ADDRESS");
+        if (toRegister) {
+            addAddressToSet(key, agent, true);
```

Comment:
Do we need to maintain a list for these agents? This currently makes things quite a bit more expensive. Instead of a fixed 20,000 gas cost to register an address using just a map, the list will cost 80,000 for the first address, and 50,000 for the following addresses. So 30,000 gas to 60,000 gas more than just a simple map. At current gas prices an increase in costs in the order of $1.

Without maintaining the list the cost is still 20,000 gas higher for the first address and 5,000 gas higher for the following addresses compared to just a map.

Reply:
Ok, I'll disable the list-query.

Reply:
Still up to 2x the cost just to be able to query how many agents someone has registered, but I guess pretty reasonable now.

Reply:
Yes, but these queries are not frequent I guess.